### PR TITLE
Moved parallel docs out of helpDB, updated and improved them

### PR DESCRIPTION
**Pull request by kshyatt**
Moved a bunch of docstrings inline. Changed things to reflect
the code that's actually in `multi.jl`. Documented `RemoteException`.

cc: @amitmurthy (again, sorry!)

### CODE REVIEW

**Comment by kshyatt** [comment]
Mac failure seems unrelated. Anybody have thoughts on this?

File: base/multi.jl [hunk 1]
```diff
@@ -368,7 +368,19 @@ const LPROC = LocalProcess()
 
 const HDR_VERSION_LEN=16
 const HDR_COOKIE_LEN=16
+
+"""
+    Base.cluster_cookie() -> cookie
+
+Returns the cluster cookie, by default the `LocalProcess` cookie.
```
Comment (amitmurthy) [inline base/multi.jl]: Reference to `LocalProcess` can be removed. It is not really mentioned in the docs anywhere and should also be probably unexported - not really relevant/useful in user code.

File: base/multi.jl [hunk 2]
```diff
@@ -650,12 +720,47 @@ function finalize_ref(r::AbstractRemoteRef)
...
+
+Create a `Future` on the local machine.
+"""
 Future(w::LocalProcess) = Future(w.id)
+
+"""
+    Future(w::Worker)
+
+Create a `Future` on the (possibly remote) worker `w`.
+"""
```
Comment (amitmurthy) [inline base/multi.jl]: `Worker` is not an exported type.

File: base/multi.jl [hunk 3]
```diff
@@ -666,7 +771,21 @@ end
 hash(r::AbstractRemoteRef, h::UInt) = hash(r.whence, hash(r.id, h))
 ==(r::AbstractRemoteRef, s::AbstractRemoteRef) = (r.whence==s.whence && r.id==s.id)
 
+"""
+    Base.remoteref_id(r::AbstractRemoteRef) -> (whence, id)
```
Comment (amitmurthy) [inline base/multi.jl]: This now returns an `RRID` type.

File: base/multi.jl [hunk 4]
```diff
@@ -832,6 +973,14 @@ type RemoteException <: Exception
...
 end
 
+"""
+    RemoteException(captured)
+
+Contains the `CapturedException` `captured`, which was
+generated on a worker and can be passed back to the first
+process as a result of `pmap` or other calls to workers
+(e.g. `remotecall_fetch`).
+"""
```
Comment (amitmurthy) [inline base/multi.jl]: Maybe reword along the lines of:

```
Exceptions  on remote computations are captured and rethrown locally.  A `RemoteException`
wraps the pid of the worker and a captured exception. A `CapturedException` captures the
 remote exception and a serializable form of the call stack when the exception was raised.

```

File: base/multi.jl [hunk 5]
```diff
@@ -871,27 +1020,61 @@ function local_remotecall_thunk(f, args, kwargs)
     return ()->f(args...; kwargs...)
 end
 
+"""
+    remotecall(f, w::LocalProcess, args...; kwargs...) -> Future
```
Comment (amitmurthy) [inline base/multi.jl]: This form of the API is for internal use. `Worker` is not exported and `LocalProcess` should be unexported.

Comment (amitmurthy) [inline base/multi.jl]: Same for all calls which refer to `Worker` and `LocalProcess`

**Comment by amitmurthy** [comment]
Thanks for doing this.

File: base/channels.jl [hunk 6]
```diff
@@ -59,6 +78,12 @@ end
 
 shift!(c::Channel) = take!(c)
 
+"""
+    isready(c::Channel)
```
Comment (tkelman) [inline base/channels.jl]: this isn't in the rst?

File: base/multi.jl (hunk 2 above)
Comment (tkelman) [inline base/multi.jl]: this signature isn't in the rst so probably fine as is?

Comment (amitmurthy) [inline base/multi.jl]: Will still show in the REPL help mode, right?

File: base/multi.jl [hunk 7]
```diff
@@ -666,7 +771,21 @@ end
 hash(r::AbstractRemoteRef, h::UInt) = hash(r.whence, hash(r.id, h))
 ==(r::AbstractRemoteRef, s::AbstractRemoteRef) = (r.whence==s.whence && r.id==s.id)
 
+"""
+    Base.remoteref_id(r::AbstractRemoteRef) -> (whence, id)
+
+A low-level API which returns the unique identifying tuple for a remote reference. A
+reference id is a tuple of two elements - `pid` where the reference was created from and a
+one-up number from that node.
```
Comment (tkelman) [inline base/multi.jl]: what does "one-up number" mean?

File: base/multi.jl (hunk 2 above)
Comment (tkelman) [inline base/multi.jl]: yeah, but I think only if you're calling it with this signature?

Comment (amitmurthy) [inline base/multi.jl]: `help?> Future` shows all docstrings defined for `Future`.

File: base/multi.jl (hunk 7 above)
Comment (amitmurthy) [inline base/multi.jl]: Maybe change the wording to:

```
`Future`s and `RemoteChannel`s are identified by fields:

`where` - refers to the node where the underlying object/storage
referred by the reference actually exists.

`whence` - refers to the node the remote reference was created from.
 Note that this is different from the node where the underlying object 
referred to actually exists. For example calling `RemoteChannel(2)` 
from the master process would results in a `where` value of 2 and 
a `whence` value of 1.

`id` is unique across all references created from the worker specified by `whence`.

Taken together,  `whence` and `id` uniquely identify a reference across all workers.

`Base.remoteref_id` is a low-level API which returns a `Base.RRID` 
object that wraps `whence` and `id` values of a remote reference.

```

File: base/multi.jl (hunk 5 above)
Comment (kshyatt) [inline base/multi.jl]: Thanks! Will remove.

File: base/multi.jl [hunk 8]
```diff
@@ -772,19 +761,34 @@ hash(r::AbstractRemoteRef, h::UInt) = hash(r.whence, hash(r.id, h))
 ==(r::AbstractRemoteRef, s::AbstractRemoteRef) = (r.whence==s.whence && r.id==s.id)
 
 """
-    Base.remoteref_id(r::AbstractRemoteRef) -> (whence, id)
+    Base.remoteref_id(r::AbstractRemoteRef) -> RRID
+
+`Future`s and `RemoteChannel`s are identified by fields:
+
+`where` - refers to the node where the underlying object/storage
+referred by the reference actually exists.
```
Comment (tkelman) [inline base/multi.jl]: referred to by

File: base/multi.jl [hunk 9]
```diff
@@ -772,19 +761,34 @@ hash(r::AbstractRemoteRef, h::UInt) = hash(r.whence, hash(r.id, h))
...
+
+`whence` - refers to the node the remote reference was created from.
+ Note that this is different from the node where the underlying object 
+ referred to actually exists. For example calling `RemoteChannel(2)` 
+ from the master process would results in a `where` value of 2 and
```
Comment (tkelman) [inline base/multi.jl]: would result in

File: base/multi.jl [hunk 10]
```diff
@@ -666,7 +760,36 @@ end
 hash(r::AbstractRemoteRef, h::UInt) = hash(r.whence, hash(r.id, h))
 ==(r::AbstractRemoteRef, s::AbstractRemoteRef) = (r.whence==s.whence && r.id==s.id)
 
+"""
+    Base.remoteref_id(r::AbstractRemoteRef) -> RRID
```
Comment (tkelman) [inline base/multi.jl]: should this go in the rst?

File: base/multi.jl [hunk 11]
```diff
@@ -424,15 +436,44 @@ function get_bind_addr(w::Worker)
...
 
+"""
+    procs()
+
+Returns a list of all process identifiers.
+"""
 procs() = Int[x.id for x in PGRP.workers]
+
+"""
+    procs(pid::Integer)
```
Comment (tkelman) [inline base/multi.jl]: add to rst?

File: base/multi.jl (hunk 10 above)
Comment (kshyatt) [inline base/multi.jl]: It doesn't seem to be referenced in the manual at all, or anywhere else in `stdlib`. Might be ok to keep the docstring only?

Comment (tkelman) [inline base/multi.jl]: it's reference by the `channel_from_id` docstring

Comment (amitmurthy) [inline base/multi.jl]: They were added in https://github.com/JuliaLang/julia/pull/14053 . Useful when serializing/deserializing only identities of remote references across workers. Both shared arrays and distributed arrays use these functions to avoid the overhead of our distributed gc in certain scenarios,

File: doc/stdlib/parallel.rst [hunk 12]
```diff
@@ -299,8 +299,6 @@ General Parallel Computing Support
 
    Called by cluster managers using custom transports. It should be called when the custom transport implementation receives the first message from a remote worker. The custom transport must manage a logical connection to the remote worker and provide two ``IO`` objects, one for incoming messages and the other for messages addressed to the remote worker. If ``incoming`` is ``true``\ , the remote peer initiated the connection. Whichever of the pair initiates the connection sends the cluster cookie and its Julia version number to perform the authentication handshake.
 
-.. function:: Future()
```
Comment (tkelman) [inline doc/stdlib/parallel.rst]: where did this go?

Comment (kshyatt) [inline doc/stdlib/parallel.rst]: Eaten by a rebase, I think.

Comment (kshyatt) [inline doc/stdlib/parallel.rst]: Oh, actually, it's extraneous now that I have the default argument shown in the other `Future` docstring.

**Comment by kshyatt** [comment]
@tkelman thoughts? Good to go?

File: base/multi.jl (hunk 4 above)
Comment (tkelman) [inline base/multi.jl]: @kshyatt not addressed

Comment (kshyatt) [inline base/multi.jl]: Thanks!

File: base/multi.jl [hunk 13]
```diff
@@ -424,15 +436,44 @@ function get_bind_addr(w::Worker)
...
+
+Returns a list of all process identifiers visible to worker `pid`.
```
Comment (amitmurthy) [inline base/multi.jl]: Returns a list of all process identifiers on the same physical node. Specifically all workers bound to the same ip-address as `pid` are returned.

Comment (amitmurthy) [inline base/multi.jl]: This is used by shared arrays to identify the subset of workers on remote nodes.

File: base/multi.jl [hunk 14]
```diff
@@ -424,15 +436,44 @@ function get_bind_addr(w::Worker)
...
+The result depends on the topology of the workgroup.
```
Comment (amitmurthy) [inline base/multi.jl]: While the statement above is not correct, it does seem more intuitive for `procs(pid)` to return workers based on topology.

I suggest we do not document this variant of `procs` for now. We can rename the current implementation of `procs(pid)` to an internal `procs_at_host(pid)` and change `procs(pid)` to return pids  based on topology.

File: doc/manual/parallel-computing.rst [hunk 15]
```diff
@@ -858,10 +858,10 @@ Custom cluster managers would typically specify only ``io`` or ``host`` / ``port
 - ``count``, ``exename`` and ``exeflags`` are relevant for launching additional workers from a worker.
   For example, a cluster manager may launch a single worker per node, and use that to launch
   additional workers.
-    - ``count`` with an integer value ``n`` will launch a total of ``n`` workers.
-    - ``count`` with a value of ``:auto`` will launch as many workers as cores on that machine.
-    - ``exename`` is the name of the ``julia`` executable including the full path.
-    - ``exeflags`` should be set to the required command line arguments for new workers.
+  - ``count`` with an integer value ``n`` will launch a total of ``n`` workers.
```
Comment (tkelman) [inline doc/manual/parallel-computing.rst]: needs rebase because of https://github.com/JuliaLang/julia/pull/17761, sorry - the newline that @MichaelHatherly added before this nested list might be necessary

File: doc/manual/parallel-computing.rst [hunk 16]
```diff
@@ -121,7 +121,7 @@ own such constructs.)
 
 An important thing to remember is that, once fetched, a :class:`Future` will cache its value
 locally. Further :func:`fetch` calls do not entail a network hop. Once all referencing
-:class:`Future`\ s have fetched, the remote stored value is deleted.
+:class:`Future` s have fetched, the remote stored value is deleted.
```
Comment (tkelman) [inline doc/manual/parallel-computing.rst]: This might have been necessary? Not sure how it changes the rendering.

File: doc/manual/parallel-computing.rst (hunk 15 above)
Comment (MichaelHatherly) [inline doc/manual/parallel-computing.rst]: Yup, the newlines above and below are needed, rst can be quite picky sometimes. (Sorry about the conflicts.)

File: doc/manual/parallel-computing.rst (hunk 16 above)
Comment (MichaelHatherly) [inline doc/manual/parallel-computing.rst]: Yes, `\` is needed if the character after a backtick isn't whitespace.

File: doc/stdlib/parallel.rst [hunk 17]
```diff
@@ -217,19 +217,31 @@ General Parallel Computing Support
 
    Returns a list of all process identifiers.
 
+.. function:: procs(pid::Integer)
+
+   .. Docstring generated from Julia source
+
+   Returns a list of all process identifiers visible to worker ``pid``\ . The result depends on the topology of the workgroup.
```
Comment (tkelman) [inline doc/stdlib/parallel.rst]: missing a `make docs` run. should squash this too, probably

**Comment by kshyatt** [comment]
Ran `make docs` and `make check-whitespace` post-squash

**Comment by kshyatt** [comment]
Thanks for all the help & review, @tkelman and @amitmurthy.
